pkg/capacity/controller_test.go

```diff
@@ -0,0 +1,1123 @@
+				}
+				if expected.uid != "" && actual.UID != expected.uid {
+					mismatches = append(mismatches, fmt.Sprintf("expected UID %s, got %s", expected.uid, actual.UID))
+				}
+				if expected.resourceVersion != "" && actual.ResourceVersion != expected.resourceVersion {
+					mismatches = append(mismatches, fmt.Sprintf("expected ResourceVersion %s, got %s", expected.resourceVersion, actual.ResourceVersion))
+				}
+				if len(mismatches) > 0 {
+					messages = append(messages, fmt.Sprintf("CSIStorageCapacity %+v:\n    %s", actual, strings.Join(mismatches, "\n    ")))
+				}
+				// Never match against the same expected capacity twice. Also, the ones that remain are dumped below.
+				expectedCapacities = append(expectedCapacities[:i], expectedCapacities[i+1:]...)
+				continue nextActual
+			}
+		}
+		messages = append(messages, fmt.Sprintf("unexpected CSIStorageCapacity %#v", actual))
+	}
+	for _, expected := range expectedCapacities {
+		messages = append(messages, fmt.Sprintf("expected CSIStorageCapacity %+v not found", expected))
+	}
+	if len(messages) > 0 {
+		return errors.New(strings.Join(messages, "\n"))
+	}
+	return nil
+}
+
+func validateCapacitiesEventually(ctx context.Context, c *Controller, clientSet *fakeclientset.Clientset, expectedCapacities []testCapacity) error {
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+	deadline, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+	var lastValidationError error
+	klog.Info("waiting for controller to catch up")
+	for {
+		select {
+		case <-ticker.C:
+			if err := process(ctx, c); err != nil {
+				return fmt.Errorf("unexpected processing error: %v", err)
+			}
+			lastValidationError = validateCapacities(ctx, clientSet, expectedCapacities)
+			if lastValidationError == nil {
+				return nil
+			}
+		case <-deadline.Done():
+			return fmt.Errorf("timed out waiting for controller, last unexpected state:\n%v", lastValidationError)
+		}
+	}
+}
+
+// createCSIStorageCapacityReactor implements the logic required for the GenerateName and UID fields to work when using
+// the fake client. Add it with client.PrependReactor to your fake client.
+func createCSIStorageCapacityReactor() func(action ktesting.Action) (handled bool, ret runtime.Object, err error) {
+	var uidCounter int
+	var mutex sync.Mutex
+	return func(action ktesting.Action) (handled bool, ret runtime.Object, err error) {
+		s := action.(ktesting.CreateAction).GetObject().(*storagev1alpha1.CSIStorageCapacity)
+		if s.Name == "" && s.GenerateName != "" {
+			s.Name = fmt.Sprintf("%s-%s", s.GenerateName, krand.String(16))
+		}
+		if s.UID == "" {
+			mutex.Lock()
+			defer mutex.Unlock()
+			uidCounter++
+			s.UID = types.UID(fmt.Sprintf("CSISC-UID-%d", uidCounter))
+		}
+		s.ResourceVersion = csiscRev + "0"
+		return false, nil, nil
+	}
+}
+
+// updateCSIStorageCapacityReactor implements the logic required for the GenerateName and UID fields to work when using
```

Comment:
It manages `ResourceVersion`, not UID nor GenerateName.

Reply:
Fixed.